File: packages/drivekit_core/flutter_drivekit_core_ios/ios/Classes/PigeonMapper.swift
import DriveKitCoreModule

extension PigeonDeleteAccountStatus {
    init(from status: DeleteAccountStatus) {
        switch status {
        case .success:
            self = .success
        case .failedToDelete:
            self = .failedToDelete
        case .forbidden:
            self = .forbidden
        @unknown default:
            fatalError()
        }
    }
}
extension PigeonRequestError {
    init(from status: RequestError) {
        switch status {
        case .wrongUrl:
            self = .wrongUrl
        case .noNetwork:
            self = .noNetwork
        case .unauthenticated:
            self = .unauthenticated
        case .forbidden:
            self = .forbidden
        case .serverError:
            self = .serverError
        case .clientError:
            self = .clientError
        case .limitReached:
            self = .limitReached
        case .unknownError:
            self = .unknownError
        @unknown default:
            fatalError()
        }
    }
}

extension PigeonUpdateUserIdStatus {
    init(from status: UpdateUserIdStatus) {
        switch status {
        case .updated:
            self = .updated
        case .failedToUpdate:
            self = .failedToUpdate
        case .invalidUserId:
            self = .invalidUserId
        case .alreadyUsed:
            self = .alreadyUsed
        case .savedForRepost:
            self = .savedForRepost
        @unknown default:
            fatalError()
        }
    }
}

extension PigeonBackgroundFetchStatus {
    init(from status: DriveKitCoreModule.DriveKitBackgroundFetchStatus) {
        switch status {
        case .started:
            self = .started
        case .completed:
            self = .completed
        @unknown default:
            fatalError()
        }
    }
}

extension PigeonDeviceConfigurationEvent {
    init(from event: DKDeviceConfigurationEvent) {
        switch event.type {
        case .locationPermission:
                if event.isValid {
                    self = .locationPermissionValid
                } else {
                    self = .locationPermissionInvalid
                }
        case .locationSensor:
                if event.isValid {
                    self = .locationSensorValid
                } else {
                    self = .locationSensorInvalid
                }
        case .activityPermission:
                if event.isValid {
                    self = .activityPermissionValid
                } else {
                    self = .activityPermissionInvalid
                }
        case .bluetoothPermission:
                if event.isValid {
                    self = .bluetoothPermissionValid
                } else {
                    self = .bluetoothPermissionInvalid
                }
        case .notificationPermission:
                if event.isValid {
                    self = .notificationPermissionValid
                } else {
                    self = .notificationPermissionInvalid
                }
        case .bluetoothSensor:
                if event.isValid {
                    self = .bluetoothSensorValid
                } else {
                    self = .bluetoothSensorInvalid
                }
        case .lowPowerMode:
                if event.isValid {
                    self = .lowPowerModeValid
                } else {
                    self = .lowPowerModeInvalid
                }
        @unknown default:
                fatalError()
        }
    }
}

File: packages/drivekit_core/flutter_drivekit_core_android/android/src/main/kotlin/com/drivequant/drivekit/flutter/core/mapper/PigeonMapper.kt
package com.drivequant.drivekit.flutter.core.mapper

import com.drivequant.drivekit.core.SynchronizationType
import com.drivequant.drivekit.core.deviceconfiguration.DKDeviceConfigurationEvent
import com.drivequant.drivekit.core.driver.UpdateUserIdStatus
import com.drivequant.drivekit.core.driver.UserInfo
import com.drivequant.drivekit.core.driver.UserInfoGetStatus
import com.drivequant.drivekit.core.driver.deletion.DeleteAccountStatus
import com.drivequant.drivekit.core.networking.RequestError
import com.drivequant.drivekit.flutter.core.PigeonDeleteAccountStatus
import com.drivequant.drivekit.flutter.core.PigeonDeviceConfigurationEvent
import com.drivequant.drivekit.flutter.core.PigeonRequestError
import com.drivequant.drivekit.flutter.core.PigeonSynchronizationType
import com.drivequant.drivekit.flutter.core.PigeonUpdateUserIdStatus
import com.drivequant.drivekit.flutter.core.PigeonUserInfo
import com.drivequant.drivekit.flutter.core.PigeonUserInfoSyncStatus

internal object PigeonMapper {

    fun UpdateUserIdStatus.toPigeon(): PigeonUpdateUserIdStatus = when (this) {
        UpdateUserIdStatus.UPDATED -> PigeonUpdateUserIdStatus.UPDATED
        UpdateUserIdStatus.FAILED_TO_UPDATE -> PigeonUpdateUserIdStatus.FAILED_TO_UPDATE
        UpdateUserIdStatus.INVALID_USER_ID -> PigeonUpdateUserIdStatus.INVALID_USER_ID
        UpdateUserIdStatus.ALREADY_USED -> PigeonUpdateUserIdStatus.ALREADY_USED
        UpdateUserIdStatus.SAVED_FOR_REPOST -> PigeonUpdateUserIdStatus.SAVED_FOR_REPOST
    }

    fun RequestError.toPigeon(): PigeonRequestError = when (this) {
        RequestError.NO_NETWORK -> PigeonRequestError.NO_NETWORK
        RequestError.UNAUTHENTICATED -> PigeonRequestError.UNAUTHENTICATED
        RequestError.FORBIDDEN -> PigeonRequestError.FORBIDDEN
        RequestError.SERVER_ERROR -> PigeonRequestError.SERVER_ERROR
        RequestError.CLIENT_ERROR -> PigeonRequestError.CLIENT_ERROR
        RequestError.UNKNOWN_ERROR -> PigeonRequestError.UNKNOWN_ERROR
        RequestError.LIMIT_REACHED -> PigeonRequestError.LIMIT_REACHED
    }

    fun DeleteAccountStatus.toPigeon(): PigeonDeleteAccountStatus = when (this) {
        DeleteAccountStatus.SUCCESS -> PigeonDeleteAccountStatus.SUCCESS
        DeleteAccountStatus.FAILED_TO_DELETE -> PigeonDeleteAccountStatus.FAILED_TO_DELETE
        DeleteAccountStatus.FORBIDDEN -> PigeonDeleteAccountStatus.FORBIDDEN
    }

    fun DKDeviceConfigurationEvent.toPigeon(): PigeonDeviceConfigurationEvent = when (this) {
        is DKDeviceConfigurationEvent.ActivityPermission -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.ACTIVITY_PERMISSION_VALID
            } else {
                PigeonDeviceConfigurationEvent.ACTIVITY_PERMISSION_INVALID
            }
        }
        is DKDeviceConfigurationEvent.AppBatteryOptimisation -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.APP_BATTERY_OPTIMISATION_VALID
            } else {
                PigeonDeviceConfigurationEvent.APP_BATTERY_OPTIMISATION_INVALID
            }
        }
        is DKDeviceConfigurationEvent.AutoResetPermission -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.AUTO_RESET_PERMISSION_VALID
            } else {
                PigeonDeviceConfigurationEvent.AUTO_RESET_PERMISSION_INVALID
            }
        }
        is DKDeviceConfigurationEvent.BluetoothSensor -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.BLUETOOTH_SENSOR_VALID
            } else {
                PigeonDeviceConfigurationEvent.BLUETOOTH_SENSOR_INVALID
            }
        }
        is DKDeviceConfigurationEvent.LocationPermission -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.LOCATION_PERMISSION_VALID
            } else {
                PigeonDeviceConfigurationEvent.LOCATION_PERMISSION_INVALID
            }
        }
        is DKDeviceConfigurationEvent.LocationSensor -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.LOCATION_SENSOR_VALID
            } else {
                PigeonDeviceConfigurationEvent.LOCATION_SENSOR_INVALID
            }
        }
        is DKDeviceConfigurationEvent.NearbyDevicesPermission -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.NEARBY_DEVICE_PERMISSION_VALID
            } else {
                PigeonDeviceConfigurationEvent.NEARBY_DEVICE_PERMISSION_INVALID
            }
        }
        is DKDeviceConfigurationEvent.NotificationPermission -> {
            if (this.isValid) {
                PigeonDeviceConfigurationEvent.NOTIFICATION_PERMISSION_VALID
            } else {
                PigeonDeviceConfigurationEvent.NOTIFICATION_PERMISSION_INVALID
            }
        }
    }

    fun UserInfoGetStatus.toPigeon(): PigeonUserInfoSyncStatus = when (this) {
        UserInfoGetStatus.SUCCESS -> PigeonUserInfoSyncStatus.SUCCESS
        UserInfoGetStatus.CACHE_DATA_ONLY -> PigeonUserInfoSyncStatus.CACHE_DATA_ONLY
        UserInfoGetStatus.FAILED_TO_SYNC_USER_INFO_CACHE_ONLY -> PigeonUserInfoSyncStatus.FAILED_TO_SYNC_USER_INFO_CACHE_ONLY
    }

    fun UserInfo?.toPigeon(): PigeonUserInfo? = this?.let {
        PigeonUserInfo(this.firstname, this.lastname, this.pseudo)
    }

    fun PigeonSynchronizationType.toModel(): SynchronizationType = when (this) {
        PigeonSynchronizationType.DEFAULT_SYNC -> SynchronizationType.DEFAULT
        PigeonSynchronizationType.CACHE -> SynchronizationType.CACHE
    }
}
